### DISCUSSION TITLE
Issue when async streaming

### DISCUSSION THREAD

**Discussion by generousjj** [tweepy/tweepy#1629] [Questions] [answered]
Sorry for messaging here, usually I'm the one answering lots of questions on discord.

I have a stream code that includes:
`class Listener(Stream):`
	`async def on_data(self, data):`

I get the following issue when running:
`/opt/virtualenvs/python3/lib/python3.8/site-packages/tweepy/streaming.py:133: RuntimeWarning: coroutine 'Listener.on_data' was never awaited`
  `self.on_data(line)`
`RuntimeWarning: Enable tracemalloc to get the object allocation traceback`

What can I do to fix this? I may have misread the async docs (since it's past 1am my time), but if I haven't this might be a module issue.

Also, how can I handle `Stream connection has errored or timed out` now that `return` values are phased out? I am using 4.0.0a.

Sorry and thank you again.

Best,
generousjj

**Comment by Harmon758** [comment, marked as answer]
[`Stream.on_data`](https://tweepy.readthedocs.io/en/latest/stream.html#tweepy.Stream.on_data) is not a coroutine. You shouldn't be defining it as one when you override it in your subclass.
If you need it to be asynchronous, consider using [`AsyncStream`](https://tweepy.readthedocs.io/en/latest/asyncstream.html) instead.

> Also, how can I handle `Stream connection has errored or timed out` now that `return` values are phased out?

Tweepy now automatically handles reconnecting when the connection errors or times out.
For handling logging of the error, you can override [`Stream.on_connection_error`](https://tweepy.readthedocs.io/en/latest/stream.html#tweepy.Stream.on_connection_error) or configure the the `tweepy` logger (See https://docs.python.org/3/howto/logging.html and https://docs.python.org/3/library/logging.html).

For code block usage, see https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks.

No need at all to apologize, but in the future, you should use [discussions](https://github.com/tweepy/tweepy/discussions/new) for questions.
